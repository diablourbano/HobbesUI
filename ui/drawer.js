import React, { Component } from 'react';
import {
  ScrollView,
  AsyncStorage,
} from 'react-native';
import { DrawerItems } from 'react-navigation';
import capitalize from 'lodash.capitalize';
import filter from 'lodash.filter';
import forEach from 'lodash.foreach';
import map from 'lodash.map';
import groupBy from 'lodash.groupby';

import { STYLEGUIDE_SYSTEM } from '../utils/constants';
import { COLORS } from '../utils/variables';
import Section from './drawerParent';
import * as styled from './styles';

storeLastRoute = async (currentRoute) => {
  try {
    await AsyncStorage.setItem('LAST_ROUTE', currentRoute);
  } catch (error) {
    // Error saving data
  }
};

retrieveLastRoute = async (currentRoute, navigation) => {
  try {
    const value = await AsyncStorage.getItem('LAST_ROUTE');
    const lastRoute = value || STYLEGUIDE_SYSTEM;

    if (currentRoute !== lastRoute) {
      navigation.navigate({ routeName: lastRoute })
    }
  } catch (error) {
  }
};

function ItemComponent(props) {
  return (
    <styled.Item
      hasPadding={props.hasPadding}
      onPress={props.onPress}
    >
      <styled.ItemList
        isActive={props.isActive}
      >
        {props.title} {props.isActive ? 'â€¢' : ''}
      </styled.ItemList>
    </styled.Item>
  );
}

function ComponentsGroup(hierarchy, props, isFiltering) {
  const {
    activeItemKey: currentRoute
  } = props;

  return map(hierarchy, (parentGroup, parentName) => (
    <Section
      key={`parent-item-${parentName}`}
      shouldCollapse={!isFiltering}
      title={capitalize(parentName)}
    >

      {map(parentGroup, (componentItem, groupName) => (
        <Section
          key={`group-item-${groupName}`}
          title={groupName}
          sectionColor={COLORS.mediumGray}
          sectionHeight={30}
          sectionFontSize={18}
          levelPosition={2}
          shouldCollapse={!isFiltering}
        >
          {map(componentItem, ({ key, title, routeName }) => (
            <ItemComponent
              hasPadding
              key={`view-item-${key}`}
              isActive={currentRoute === key}
              title={title}
              onPress={() => {
                storeLastRoute(routeName);
                props.navigation.navigate({ routeName })
              }}
            />
          ))}
        </Section>
      ))}
    </Section>
  ));
}

export class CustomDrawer extends Component {
  state = {
    filterValue: null,
    filteredComps:[],
    isFiltering: false,
  };

  componentDidMount() {
    const { activeItemKey: currentRoute, navigation } = this.props;
    retrieveLastRoute(currentRoute, navigation);
  }

  filterStyleguides = (text) => {
    const { customItems } = this.props;
    const [Welcome, ...componentItems] = customItems;

    const isFiltering = !!text;

    const filterRegex = new RegExp(text, 'i');
    const filteredComps = isFiltering ?
      filter(componentItems, (item) =>
        filterRegex.test(item.title) || filterRegex.test(item.group)
      ) :
      [];

    this.setState({
      filterValue: text,
      filteredComps,
      isFiltering,
    });
  };

  render() {
    const { isFiltering, filterValue, filteredComps } = this.state;
    const {
      customItems,
      activeItemKey: currentRoute,
      navigation,
    } = this.props;

    const [Welcome, ...componentItems] = customItems;

    const hierarchy = groupBy(isFiltering ? filteredComps : componentItems, 'parent');
    forEach(hierarchy, (group, parent) => {
      hierarchy[parent] = groupBy(group, 'group');
    });

    const isActive = currentRoute === Welcome.key;

    return (
      <ScrollView>
        <styled.SafeView
          forceInset={{ top: 'always', horizontal: 'never' }}
        >
          <styled.FilterComps
            placeholder="filter..."
            value={filterValue}
            onChangeText={this.filterStyleguides}
          />

          {!isFiltering && (
            <ItemComponent
              isActive={currentRoute === Welcome.key}
              title={Welcome.title}
              onPress={() => {
                storeLastRoute(Welcome.routeName);
                navigation.navigate({ routeName: Welcome.routeName })
              }}
            />
          )}
          {ComponentsGroup(hierarchy, this.props, isFiltering)}
        </styled.SafeView>
      </ScrollView>
    );
  }
}
