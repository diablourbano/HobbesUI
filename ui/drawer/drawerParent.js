import React, { Component } from 'react';
import { COLORS } from '../../utils/variables';
import * as styled from '../styles';

class DrawerSection extends Component {
  state = {
    isCollapsed: true,
  };

  componentDidUpdate(prevProps) {
    const { shouldCollapse: prevShouldCollapse } = prevProps;
    const { shouldCollapse } = this.props;

    if (shouldCollapse !== prevShouldCollapse) {
      this.setCollapsedTo(shouldCollapse);
    }
  }

  toggleCollapsed = () => {
    const { isCollapsed: wasCollapsed } = this.state;
    this.setCollapsedTo(!wasCollapsed);
  };

  setCollapsedTo(state) {
    this.setState({ isCollapsed: state });
  }

  render() {
    const {
      children,
      title,
      sectionColor,
      sectionHeight,
      sectionFontSize,
      levelPosition,
    } = this.props;
    const { isCollapsed } = this.state;

    return (
      <styled.Parent
        isCollapsed={isCollapsed}
        sectionHeight={sectionHeight}
        levelPosition={levelPosition}
        sectionColor={sectionColor}
      >
        <styled.ParentSection
          sectionColor={sectionColor}
          sectionHeight={sectionHeight}
          onPress={this.toggleCollapsed}
        >
          <styled.ParentName
            sectionFontSize={sectionFontSize}
          >
            {title}
          </styled.ParentName>
        </styled.ParentSection>

        {children}
      </styled.Parent>
    );
  }
}

DrawerSection.defaultProps = {
  sectionColor: COLORS.darkGray,
  sectionHeight: 32,
  sectionFontSize: 20,
  levelPosition: 1,
};

export default DrawerSection;
