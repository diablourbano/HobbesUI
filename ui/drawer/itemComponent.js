import React from 'react';
import * as styled from '../styles';

export const ItemComponent = (props) => {
  const { hasPadding, onPress, isActive = false, title } = props;

  return !title || !onPress ? null :(
    <styled.Item
      hasPadding={hasPadding}
      onPress={onPress}
    >
      <styled.ItemList
        isActive={isActive}
      >
        {title} {isActive ? 'â€¢' : ''}
      </styled.ItemList>
    </styled.Item>
  );
}
