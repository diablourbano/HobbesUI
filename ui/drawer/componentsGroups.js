import React from 'react';
import { capitalize, findIndex, forEach, map } from 'lodash';
import { storeLastRoute } from '../../utils/routeStore';
import { COLORS } from '../../utils/variables';
import { ItemComponent } from './itemComponent';
import Section from './drawerParent';

export const ComponentsGroup = (props) => {
  const { hierarchy, isFiltering, navigation, currentRoute } = props;

  return !hierarchy ? null : map(hierarchy, (parentGroup, parentName) => {
    let intoCurrentRoute = false;
    const parentGroupArr = Object.values(parentGroup);

    forEach(parentGroupArr, (group) => {
      if (findIndex(group, { routeName: currentRoute }) > -1) {
        intoCurrentRoute = true;
        return false;
      }
      return true;
    });

    return (
      <Section
        key={`parent-item-${parentName}`}
        shouldCollapse={!isFiltering && !intoCurrentRoute}
        title={capitalize(parentName)}
      >

        {map(parentGroup, (componentItem, groupName) => {
          const intoCurrentGroup = findIndex(componentItem,
            { routeName: currentRoute }) > -1;

          return (
            <Section
              key={`group-item-${groupName}`}
              title={groupName}
              sectionColor={COLORS.mediumGray}
              sectionHeight={30}
              sectionFontSize={18}
              levelPosition={2}
              shouldCollapse={!isFiltering && !intoCurrentGroup}
            >
              {map(componentItem, ({ key, title, routeName }) => (
                <ItemComponent
                  hasPadding
                  key={`view-item-${key}`}
                  isActive={currentRoute === key}
                  title={title}
                  onPress={() => {
                    storeLastRoute(routeName);
                    navigation.navigate(routeName);
                  }}
                />
              ))}
            </Section>
          );
        })}
      </Section>
    );
  });
};
