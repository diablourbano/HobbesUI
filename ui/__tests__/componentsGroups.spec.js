import React from 'react';
import { shallow } from 'enzyme';
import { ComponentsGroup } from '../drawer/componentsGroups';

describe('ComponentsGroup snapshots', () => {
  it('should return null if hierarchy is undefined', () => {
    const component = shallow((
      <ComponentsGroup />
    ));
    expect(component.isEmptyRender()).toEqual(true);
  });

  it('should return [] if hierarchy is empty', () => {
    const component = shallow((
      <ComponentsGroup hierarchy={[]} />
    ));
    expect(component).toMatchSnapshot();
  });

  // it('should', () => {
  //   const component = shallow(
  //   );

  //   expect(component).toMatchSnapshot();
  // });
});
