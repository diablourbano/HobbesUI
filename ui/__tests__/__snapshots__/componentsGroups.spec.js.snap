// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentsGroup snapshots should return [] if hierarchy is empty 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ComponentsGroup
    hierarchy={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): undefined,
  Symbol(enzyme.__nodes__): Array [],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
